File: ts/code/canFormArray.ts
/**
 * @author heart
 * @description
 * @Date 2022-09-22
 * @see https://leetcode.cn/problems/check-array-formation-through-concatenation/solution/
 */
function canFormArray(arr: number[], pieces: number[][]): boolean {
  for (let i = 0; i < arr.length; i++) {
    const index = pieces.findIndex((val) => val.includes(arr[i]))
    if (index === -1) return false
    const [list] = pieces.splice(index, 1)
    for (let j = 0; j < list.length; j++) {
      if (list[j] !== arr[i + j]) {
        return false
      }
    }
    i = i + list.length - 1
  }
  return true
}

File: go/src/canFormArray.go
// Package al @see https://leetcode.cn/problems/check-array-formation-through-concatenation/
package al

func canFormArray(arr []int, pieces [][]int) bool {
	m := make(map[int]int, len(pieces))
	for index, list := range pieces {
		// 存储第一位的下标
		m[list[0]] = index
	}
	for i := 0; i < len(arr); {
		j, ok := m[arr[i]]
		if !ok {
			return false
		}
		for _, x := range pieces[j] {
			if arr[i] != x {
				return false
			}
			i++
		}
	}
	return true
}
